model/map/src/main/java/org/keycloak/models/map/realm/MapRealmProvider.java

```diff
@@ -0,0 +1,469 @@
+        return entity == null ? null : entityToAdapterFunc().apply(entity);
+    }
+
+    @Override
+    public RealmModel getRealmByName(String name) {
+        if (name == null) return null;
+
+        LOG.tracef("getRealmByName(%s)%s", name, getShortStackTrace());
+
+        ModelCriteriaBuilder<RealmModel> mcb = realmStore.createCriteriaBuilder()
+                .compare(SearchableFields.NAME, Operator.EQ, name);
+
+        String realmId = tx.getUpdatedNotRemoved(mcb)
+                .findFirst()
+                .map(MapRealmEntity::getId)
+                .orElse(null);
+        //we need to go via session.realms() not to bypass cache
+        return realmId == null ? null : session.realms().getRealm(realmId);
+    }
+
+    @Override
+    public Stream<RealmModel> getRealmsStream() {
+        return tx.getUpdatedNotRemoved(realmStore.createCriteriaBuilder())
+                .map(entityToAdapterFunc())
+                .sorted(RealmModel.COMPARE_BY_NAME);
+    }
+
+    @Override
+    public Stream<RealmModel> getRealmsWithProviderTypeStream(Class<?> type) {
+        return StreamUtils.leftInnerJoinStream(tx.getUpdatedNotRemoved(realmStore.createCriteriaBuilder()), MapRealmEntity::getComponents) // create pair of realm and its components
+                .filter(pair -> Objects.equals(type.toString(), pair.getV().getProviderType())) // filter components based on type
+                .map(Pair::getK) // get remaining realms
+                .map(entityToAdapterFunc())
+                .sorted(RealmModel.COMPARE_BY_NAME);
+    }
+
+    @Override
+    public boolean removeRealm(String id) {
+        LOG.tracef("removeRealm(%s)%s", id, getShortStackTrace());
+
+        RealmModel realm = getRealm(id);
+
+        session.users().preRemove(realm);
+        session.clients().removeClients(realm);
+        session.clientScopes().removeClientScopes(realm);
+        session.roles().removeRoles(realm);
+        realm.getTopLevelGroupsStream().forEach(realm::removeGroup);
+
+        // TODO: Sending an event should be extracted to store layer
+        session.getKeycloakSessionFactory().publish(new RealmModel.RealmRemovedEvent() {
+            @Override
+            public RealmModel getRealm() {
+                return realm;
+            }
+
+            @Override
+            public KeycloakSession getKeycloakSession() {
+                return session;
+            }
+        });
+        // TODO: ^^^^^^^ Up to here
+
+        tx.delete(id);
+        return true;
+    }
+
+    @Override
+    public void removeExpiredClientInitialAccess() {
+        ModelCriteriaBuilder<RealmModel> mcb = realmStore.createCriteriaBuilder()
+                .compare(SearchableFields.CLIENT_INITIAL_ACCESS, Operator.EQ, true);
+
+        tx.getUpdatedNotRemoved(mcb)
+                .forEach(MapRealmEntity::removeExpiredClientInitialAccesses);
+    }
+
+    @Override
+    public void saveLocalizationText(RealmModel realm, String locale, String key, String text) {
```

Comment:
This method is not touching store, it just uses adapter in some way. I remember seeing something like this already in some other provider (maybe in group), should we move methods like that to adapter? Probably in the cleanup task, here we should just add TODO if we want to do it. WDYT?

Reply:
It sounds good, would you like me to add `//TODO move the following method to adapter` comment?

Reply:
I think this is something we can discuss in the meeting as well.

Reply:
yes, please add the TODO here